text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+
+We probably don't want to be doing a PayJoin transaction when we are withdrawing from an exchange to showing them one of our inputs. To avoid this, we can generate a `SlatepackAddress` from a configuration option that has 0% chance to do a PayJoin receive transaction.
+
+### PayJoin transactions
+
+PayJoin transactions when mixed with regular transaction introduce some uncertainty to the inputs side of a transaction. When Alice is sending coins to Bob, Bob can decide to contribute his own input to a transaction which makes it harder for chain analysis tools to determine who is the owner of an input - sender's inputs become mixed with the receiver's inputs. So it contributes to the sender's privacy, but hurts a bit the receiver's privacy because they show one of their inputs. The only party that knows the receiver contributed an input is the sender though.
+
+Currently, PayJoin transactions are cheaper than regular transactions because they have more inputs. This also means that there are less outputs created which helps to reduce unnecessary chain growth. A 2-2 PayJoin transaction does not create a new output as opposed to a regular 1-2 transaction, which means that it can be thought of as an replacement of the value of the current output with a new value which does not increase the UTXO set size. Another side effect of PayJoin transactions is that backwards chain analysis becomes much harder to follow than it is right now.
+
+### Play attack protection
+
+The difference between a Play and Replay attacks are that in a Play attack, the transaction never lands on the chain. This allows some funny scenarios where Alice wants to pay Bob, but after constructing the transaction and broadcasting it, the transaction does not land on the chain for some reason. Let's say that Alice and Bob recreate a new transaction and Alice uses different outputs. This second transaction goes through and both Alice and Bob are happy. However, if Bob saw the first transaction, if it becomes valid at some point, it is possible for him to 'play' it or broadcast it on the chain after the second transaction was already done. This way, Bob receives his payment twice. This means that a Play attack attacks the sender by tricking them into signing a transaction multiple times. Transactions that don't make it to the chain should _always_ be cancelled by the user before sending a new transaction.
+
+To protect against this, a user should have an option to cancel a transaction which would label the sender's inputs as `Must use` in the next transaction. The reason why the sender would want to reuse the inputs is to make the previous transaction invalid. So now, only 1 of the two transactions that have been signed can make it to the chain. If the transaction failed to get on the blockchain again and Bob 'gave up', it should be cancelled and an immediate self-spend should be done to prevent giving Bob the possibility of playing the transaction after.

```

Comment:
This double-spending cancel option must be used with care. We should not reuse an input of the canceled tx as an input for a receiving payjoin, as it exposes the sender to the possibility of unexpected failure (when Bob has the canceled tx confirm after all).
Similarly, such failure is possible if we reuse an input for a spending transaction. So in that case we must watch the chain until the new spend confirms. If it fails due to the original tx having been broadcast/confirmed, then we must be prepared to  replace the now bad input by new ones, adjust the offset, and rebroadcast the send.
Clearly this is something only an expert user should be allowed to do.
For casual users, cancel should always lead to an immediate self-spend.

Reply:
Ugh nasty. Will correct to immediate self-spend. Perhaps we might also want to think of a `resend` option that would reuse the same inputs to recreate the transaction with the same receiver, but this can be thought about later.